integration-cli/docker_cli_pull_test.go

```diff
@@ -39,30 +40,68 @@ func (s *DockerHubPullSuite) TestPullFromCentralRegistry(c *check.C) {
 // combinations of implicit tag and library prefix.
 func (s *DockerHubPullSuite) TestPullNonExistingImage(c *check.C) {
 	testRequires(c, DaemonIsLinux)
-	for _, e := range []struct {
+
+	type entry struct {
 		Repo  string
 		Alias string
-	}{
+	}
+
+	entries := []entry{
 		{"library/asdfasdf", "asdfasdf:foobar"},
 		{"library/asdfasdf", "library/asdfasdf:foobar"},
 		{"library/asdfasdf", "asdfasdf"},
 		{"library/asdfasdf", "asdfasdf:latest"},
 		{"library/asdfasdf", "library/asdfasdf"},
 		{"library/asdfasdf", "library/asdfasdf:latest"},
-	} {
-		out, err := s.CmdWithError("pull", e.Alias)
-		c.Assert(err, checker.NotNil, check.Commentf("expected non-zero exit status when pulling non-existing image: %s", out))
-		// Hub returns 401 rather than 404 for nonexistent repos over
-		// the v2 protocol - but we should end up falling back to v1,
-		// which does return a 404.
-		c.Assert(out, checker.Contains, fmt.Sprintf("Error: image %s not found", e.Repo), check.Commentf("expected image not found error messages"))
-
-		// pull -a on a nonexistent registry should fall back as well
+	}
+
+	// The option field indicates "-a" or not.
+	type record struct {
+		e      entry
+		option string
+		out    string
+		err    error
+	}
+
+	// Execute 'docker pull' in parallel, pass results (out, err) and
+	// necessary information ("-a" or not, and the image name) to channel.
+	var group sync.WaitGroup
```

Comment:
I think the code would be a bit simpler if instead of using a `WaitGroup`, the loop over `entries` counted the number of goroutines started, and the loop below it that reads from the channel runs that many times. That would avoid the need to use a buffered channel. But it's only a very minor simplification, so I'm fine with this as-is if you prefer it that way.